### PR TITLE
U4-6003 List View - Order By Custom Property (Fix for 7.4.3)

### PR DESCRIPTION
**Pull request by aadfPT**
Mostly the existing fix from http://issues.umbraco.org/issue/U4-6003, although commit d560223 was used to fix two other issues, that still prevented it from working.

### CODE REVIEW

**Comment by lars-erik** [comment]
Hi,

I'm also committed to get this working, but don't have the time currently. (Not HQ)  
Great that you did some work on it, but I'm fairly sure this won't get pulled in it's current state.  
You need to get some conventions down first.  
Please read through these two docs:  
https://our.umbraco.org/contribute/guidelines-for-core-contribution/  
https://our.umbraco.org/documentation/Development-Guidelines/Coding-Standards/  
One of the most important things is whitespace. It's impossible to see what you changed since the whitespacing has changed.  
Also, the commit messages could do with some more details like Andy did on the original PR.  
Looking forward to seeing the issue fixed. :)  

You could also do a pull request to Andy's branch so it's easier to follow, and one unified team effort.

**Comment by aadfPT** [comment]
Hi Lars,
which branch would that be?
Feel free to do the pull request, I have changed the whitespace of the files I edited.

**Comment by lars-erik** [comment]
Try adding https://github.com/AndyButland/Umbraco-CMS.git as a remote, check out his branch u6003b.
Then try merging in your own branch to andy's branch u6003b - and after that, merge in origin/dev-v7.
Make a PR on his fork. https://github.com/AndyButland/Umbraco-CMS

Not quite sure this is gonna work, but it should. :) You may possible have to cherry-pick or just copy from your stuff.

This is a rather tough one since the changed classes are pretty volatile, and the HQ guys need the PR to be as clean as possible to easily review it for a pull.

**Comment by aadfPT** [comment]
Done here
https://github.com/AndyButland/Umbraco-CMS/pull/1

I'm leaving it to you both from now on, and removing myself from the equation :-)

**Comment by Shazwazza** [comment]
Thanks all, i'll start reviewing this stuff on Monday, getting this all done is scheduled for this sprint.

**Comment by lars-erik** [comment]
Looks a lot easier for Shaz to review now, @aadfPT. 
Just have another look at .gitignore. Too many files for me to look into now, but seems like you got it. :+1:

**Comment by Shazwazza** [comment]
I'm re-opening this PR. The PR against Andy's commit has gone wrong, I'm pretty sure an incorrect branch was targeted -- as seen here: https://github.com/AndyButland/Umbraco-CMS/pull/1 there are 2645 files changed in that PR.

I would like to understand what this PR is.... 

* Is it a PR that includes Andy's changes in his PR with additional changes, or?
* Is it a PR that is changes required in addition to Andy's PR?

**Comment by lars-erik** [comment]
I attempted to help @aadfPT rebase Andy's on the newest master-v7, and then apply his own updates to make it compatible with 7.4 on andy's stuff. Hence the 2k commits from master-v7 since Andy wrote his stuff.

**Comment by Shazwazza** [comment]
Right, so I'd need to wait until @AndyButland pulls in that PR and re-push his branch for this PR to see all changes.

So i will assume that this particular PR is change in addition to Andy's PR. 

I'm going to look at getting Andy's PR in now after some review. Depending on how long that takes maybe Andy can accept the other changes before I get done but we'll see.

**Comment by lars-erik** [comment]
Yep. I thought you'd be able to get this. I guess you could clone @aadfPT's branch.
Andy's branch was made for 7.2 or 3, and a lot of the intefaces was changed after his changes. So the majority of stuff aadfPT should've done would be updated signatures.  
I stopped my attempt when I saw the amount of changed sigs. :)

**Comment by Shazwazza** [comment]
I'm just going to command line merge Andy's and fix everything up, then i'll command line merge this one on top and see how I go.

**Comment by aadfPT** [comment]
I did find it particularly troublesome to have to PR into a secondary branch of a thirdparty repository, which was still set to version 6 of Umbraco, but did as asked.
@Shazwazza if you review the changes commited, you'll see that all of the commits included already exist on the dev7 tree of the Main Umbraco repository. It's simply pulling them over to an out of date branch on Andy's account.
My commits are but 4 or so at the end which you can check by verifying the author.
The changes should be easy to follow, let me know if you have any questions.
If you decide to pull directly from my repository (which in my opinion would be the best option), this pull request is still valid. The reason the check failed is I had to fix a bug of a null reference that was since fixed in the main branch. Extremely simple to solve when doing conflict resolution on the merge.
Feel free to copy the changes manually / cherry pick or whatever you prefer in order to import them into Umbraco.

**Comment by Shazwazza** [comment]
No worries, once i get through fixing up Andy's i'll get yours all merged in. Thanks!

**Comment by AndyButland** [comment]
Hi - just touching base too.  Thanks @aadfPT for the PR a from a few days ago and sorry not got to it yet.  I did have a little think about it but just clicking a button to merge in 2500+ commits made me hesitate a bit!

@Shazwazza - sound like you are working through it all so just let me know what, if anything, there's further useful to do on my end.  Clearly it's been a while since the original PR but I can refresh my memory and make some changes if they are needed.  Or if you can pull from me, @aadfPT and fix things up yourself, that's all good too.

**Comment by aadfPT** [comment]
@AndyButland I won't do a pull from you simply because your fix is for the wrong version of umbraco.
For that, I would want you to first take the latest version of dev7 from the main repo, then I could pull from yours, fix and pr to yours again. But I already did that directly from the official repository hence I don't see a need.
As I said before the commits on the pull request are mostly from the main repository, so..
You can update your branch first (by pulling from dev7), and then revisit my pull request, which will then be reduced to my changes.

**Comment by Shazwazza** [comment]
All good guys, i'm sorting through this :)

**Comment by Shazwazza** [comment]
I'm closing this PR now, have got it merged in to a 7.5 targeted branch currently: 7.5-ListViewCustomFieldSorting

Wonderful work!! ... however I do have one question which I'll ask inline in the code.

File: src/Umbraco.Web/Editors/ContentController.cs [hunk 1]
```diff
@@ -179,18 +179,22 @@ public HttpResponseMessage GetNiceUrl(int id)
...
-            int pageSize = 0, 
-            string orderBy = "SortOrder", 
-            Direction orderDirection = Direction.Ascending, 
-            string filter = "")
+                int id,
+                int pageNumber = 0,  //TODO: This should be '1' as it's not the index
+                int pageSize = 0,
+                string orderBy = "SortOrder",
+                Direction orderDirection = Direction.Ascending,
+                int? orderBySystemField = 2,
```
Comment (Shazwazza) [inline src/Umbraco.Web/Editors/ContentController.cs]: @aadfPT I'm trying to figure out why this is an integer param, any ideas? The other controllers have this as a boolean so need to know why/how this is an integer.

**Comment by aadfPT** [comment]
The value is posted as a number, and for some reason it wouldn't recognize it properly when the signature received a bool.
I set it to 2 by default to know if it was even posted, then I convert to bool if it's different of 0.
It was just a quick fix to make it work. Ideally bool would work out of the box.

File: src/Umbraco.Web/Editors/ContentController.cs (hunk 1 above)
Comment (aadfPT) [inline src/Umbraco.Web/Editors/ContentController.cs]: The value is posted as a number, and for some reason it wouldn't recognize it properly when the signature received a bool.
I set it to 2 by default to know if it was even posted, then I convert to bool depending on the value.
It was just a quick fix to make it work. Ideally bool would work out of the box.

Comment (Shazwazza) [inline src/Umbraco.Web/Editors/ContentController.cs]: All good, have fixed this up now

**Pull request closed without merging**